src/typedefs.jl

```diff
@@ -2,9 +2,10 @@ const DefaultRefType = UInt32
 
 ## Pools
 
-# V is always set to CategoricalValue{T}
-# This workaround is needed since this type not defined yet
-# See JuliaLang/julia#269
+# Type params:
+# * `T` type of categorized values
+# * `R` integral type for referncing category levels
+# * `V` categorical value type that implements "categorical value" concept
```

Comment:
Instead of "concept", which isn't a standard term, I'd just say "one of the `CatValue` types", and enforce this in the signature with `<:CatValue`. That makes it clear that for now this is not a public extensible interface. If you want, call this "trait", as "concept" is pretty vague (same below).

Reply:
I'm using "concept" in C++1x sense, but you're right, probably it's better to say "categorical value trait".

In the final version of PR I would like to reduce the use of `CatValue` to `values.jl` only.
I think it's possible to make `CategoricalArray` dispatch based on `iscatvalue(C)` only without introducing any more complexity than it currently has.

Reply:
I'm afraid dispatching on a trait will be more complex than using `CatValue`, as you need to pass `iscatvalue(C)` as a parameter instead of just dispatching on the argument's types. It's fine to use `CatValue`, no need to make things more complex unless/until we have a good reason to.